### ISSUE TITLE
Heavy Tearing On Fullscreen.

### ISSUE THREAD

**Issue by Gerarama** [godotengine/godot#41946]
**Godot version:**
Godot 3.2.2 stable

**OS/device including version:**
  GTX 1070 Windows 10 1909 GLES 2,3

**Issue description:**

fullscreen has mush more(or the only mode that has) tearing when disabling Vsync.

**Steps to reproduce:**
Disable Vsync
Shake the screen on window mode and then on fullscreen and see the tearing.

**Minimal reproduction project:**

Press Space to toggle fullscreen and window mode.

[Tearing.zip](https://github.com/godotengine/godot/files/5201343/Tearing.zip)

**Comment by Calinou** [comment]
You will always get tearing with V-Sync disabled. Godot can't bypass the monitor's physical limitations :slightly_smiling_face:

**Comment by Gerarama** [comment]
>>Playing in a smaller window will make tearing less noticeable since the window will occupy a lower portion of your vision.

@Calinou 
How about maximized window?
with resulotion of 8K? shouldn't that put some tearing?
and force fps to be 60 or below?

i can't see a single tear, it only happen in fullscreen or maximized borderless.

**Comment by Gerarama** [comment]
@Calinou 

making project setting resulotion exactly my screen resulotion and putting it borderless get rid of the tearing in fullscreen like display.

How do you explain that?

**Comment by Calinou** [comment]
@Gerarama Windows gives different a process priority to applications in fullscreen mode compared to those in windowed mode. See [Fixing jitter and stutter](https://docs.godotengine.org/en/latest/tutorials/misc/jitter_stutter.html) in the documentation.

**Comment by Gerarama** [comment]
@Calinou 
The link you doesn't explain how this solution works at all.
it talks about stutter effect between windowed and fullscreen mode, which is not our problem anyway.

based on the doc a special priority get rid of the stutter.
and create tearing problem?

so doesn't getting rid of stutter result in tearing?

**Comment by Calinou** [comment]
> so doesn't getting rid of stutter result in tearing?

V-Sync prevents tearing by increasing the delay with which frames are displayed. You can't have your cake and eat it, unless you have a *variable refresh rate monitor* (NVIDIA G-Sync or AMD FreeSync). Variable refresh rate requires compatible hardware and some special setup to get it to work flawlessly, so it's not for everyone.

**Comment by TheDuriel** [comment]
After laborious testing and talking with @Gerarama in discord I can confirm that they do not understand how screen tearing happens, and what the various approaches are for fixing it. And how and when the desktop window manager of windows will enable/disable vsync settings.

Everything is working fine.

Tearing occurs in the provided project in all cases in which vsync is disabled. This means disabled within godot, and disabled within windows and or drivers.

If vsync is enabled, in godot or via the desktop window manager or drivers, then no tearing occurs.

I have a RTX2070 super, running the project at upwards of 2000 fps with vsync disabled. and a 144hz gsync _compatible_ monitor. (however, godots vsync implementation is locked to 60hz and is not compatible with gsync at this date)

TL:DR: Issue can remain closed. No issue to be found.

**Comment by TheDuriel** [comment]
To further squash this: the photos are confirmed to be taking with a camera suffering from long exposure time and long shutter speed. Meaning that images become blurry unless a steadycam or tripod is used, and that either way it will capture several frames to composite one image. Hence why its blurry/ghosting. Its entirely the cameras fault.

**Comment by Gerarama** [comment]
Having talked to @TheDuriel in discord he seems to not undertand how vsync work, and how there is no relation between the window modes and vsync being enabled or disabled, having i prove it with my testing of borderless windows.

he kept talking about fullscreen optimization and windows priority and non sense without proofs.
and he said that it can be fixed by changing some windows settings and he didn't show any.

>To further squash this: the photos are confirmed to be taking with a camera suffering from long exposure time and long shutter speed. Meaning that images become blurry unless a steadycam or tripod is used, and that either way it will capture several frames to composite one image. Hence why its blurry/ghosting. Its entirely the cameras fault.

and here is trying to make it sound stupid by blaming my camera.

even tho its So clear that the tearing is happening, but heck here is even better photos with better camera:

![119118698_315981152970830_3935544091199037463_n](https://user-images.githubusercontent.com/61591052/92747121-904a9780-f38c-11ea-8d17-dbb0fe2e2cfa.jpg)

![119126609_620969948782530_6208971030582369272_n](https://user-images.githubusercontent.com/61591052/92747139-93458800-f38c-11ea-90ee-5b549a77c966.jpg)

![119167546_322352979018458_333052287491223438_n](https://user-images.githubusercontent.com/61591052/92747159-9a6c9600-f38c-11ea-8352-2bda9487c5cc.jpg)

**Comment by darksylinc** [comment]
This ticket is pointless. VSync is the method to prevent tearing. Disabling VSync disables any guarantees regarding tearing. The picture on the screen may or may not have tearing.

> making project setting resulotion exactly my screen resulotion and putting it borderless get rid of the tearing in fullscreen like display.
> How do you explain that?

You could as well say "Shooting people in the head kills them, please fix it", and then "I just shot someone in the head and he survived, how do you explain that?"

No guarantees when VSync is off means just that: no guarantees.

Sometimes tearing is reduced because the different conditions cause a slight shift in framerate, causing the natural framerate of the monitor to align with the game's, reducing the frequency in which tearing manifests.

Other times it's because the OS is performing vsync behind your back (i.e. silently introduces latency) or changes how presentation is being made (blitting vs flipping), thus window, borderless window and exclusive fullscreen all manifest different kind of tearing behaviour or even no tearing at all.

**Comment by Gerarama** [comment]
@darksylinc 
Nope, its indeed using my project settings only when the fullscreen is enabled.
see `gl_context->set_use_vsync(video_mode.use_vsync);` in `os.windows.cpp`

You havn't added any new information to this topic,
rather then confirming that OS has something to do with this.

What this topic was all about is why does tearing only happen in fullscreen when disabling vsync, while no tearing ever happen in windowed mode.

If you can't answer that, then ignore the issue, because if i want vsync to be off in windowed mode, then its an issue and i have the right to open it.

**Comment by darksylinc** [comment]
> What this topic was all about is why does tearing only happen in fullscreen when disabling vsync, while no tearing ever happen in windowed mode.

Because in windowed mode the compositor is rendering to the screen using VSync, and often the game present to the compositor via the blitting model.

When you switch to fullscreen, it is a direct path to the monitor. No compositor (hence no implicit VSync) and the presentation model to the monitor is often _flip_ instead of _blit_.

If you suspect VSync is always enabled in windowed mode, then the framerate must be locked to 60 fps (assuming a 60hz monitor and the computer is fast enough) while in windowed mode, while uncapped in fullscreen.

If you get e.g. 1000 fps while in windowed mode, then VSync is off. Whether you get tearing or not.

**Comment by Gerarama** [comment]
finally some reasonable statement.
True that explains everything.
And it sonuds like there is no way to disable it, Because its Windows thing.

Thank you!!

**Comment by Skyqwerty11** [comment]
> **Godot version:**
> Godot 3.2.2 stable
> 
> 
> **OS/device including version:**
>   GTX 1070 Windows 10 1909 GLES 2,3
> 
> 
> **Issue description:**
> 
> fullscreen has mush more tearing then windowed mode while disabling Vsync
> 
> **Steps to reproduce:**
> Disable Vsync
> Shake the screen on window mode and then on fullscreen and see the tearing.
> Then switch fullscreen off and see the tearing
> 
> **Minimal reproduction project:**
> 
> Press Space to toggle fullscreen and window mode.
> 
> [Tearing.zip](https://github.com/godotengine/godot/files/5201343/Tearing.zip)
> 
>
